packages/puppeteer-core/src/util/decorators.ts

```diff
@@ -72,3 +72,36 @@ export function throwIfDisposed<This extends Disposed>(
     };
   };
 }
+
+export function invokeAtMostOnceForArguments(
+  target: (this: unknown, ...args: any[]) => any,
+  _: unknown
+): typeof target {
+  const cache = new WeakMap();
+  let cacheDepth = -1;
+  return function (this: unknown, ...args: unknown[]) {
+    if (cacheDepth === -1) {
+      cacheDepth = args.length;
+    }
+    if (cacheDepth !== args.length) {
+      throw new Error(
+        'Memoized method was called with the wrong number of arguments'
```

Comment:
What happens if the function has optional paramater `fn(a: object, b: object, c?: object)` and we call `fn({a},{b})` and then `fn({a},{b},{c})` this will throw right? I don't think it should.

Reply:
Yeah, it's the drawback, the number of arguments has to remain the same. I am not sure it's worth supporting dynamic arguments.

Reply:
Let's leave a comment on the decorator itself and we can merge this.